lesson08.md

```diff
@@ -0,0 +1,364 @@
+
+# Functions
+
+---
+
+### Why we need functions
+
+- to add custom functionalities to JavaScript 
+- to group code together
+- to avoid repetition 
+- to easily repeat set of instructions
+- to organise our code better
+- to hide complexity
+
+---
+
+### Example: introduce yourself
+
+```js
+let melanieIntroduction = "Hello, my name is Melanie. I come from Germany and my favuorite food is broccoli";
+let marcoIntroduction = "Hello, my name is Marco. I come from Italy and my favuorite food is cheese";
+let nataliaIntroduction = "Hello, my name is Natalia. I come from Spain and my favuorite food is lentils";
+
+console.log(melanieIntroduction);
+console.log(marcoIntroduction);
+console.log(nataliaIntroduction);
+```
+
+
+---
+
+### Example: introduce yourself
+
+```js
+function introduceYourself(name, originCountry, favouriteFood) {
+    let message = "Hello, my name is " + name + ". I come from " + originCountry + " and my favuorite food is " + favouriteFood;
+    return message;
+}
+
+let melanieIntroduction = introduceYourself("Melanie", "Germany", "broccoli");
+let marcoIntroduction = introduceYourself("Marco", "Italy", "cheese");
+let nataliaIntroduction = introduceYourself("Natalia", "Spain", "lentils");
+
+console.log(melanieIntroduction);
+console.log(marcoIntroduction);
+console.log(nataliaIntroduction);
+```
+
+---
+
+### Example: introduce yourself
+
+What important differences do you notice between the two versions?
+
+Which version do you think is better?
+
+Why?
+
+---
+
+
+### Example: introduce yourself
+
+To better understand the benefit of using functions, imagine you have to do the simple task of changing the message from using "Hello" to "Hi".
+
+With functions, you have to change the code only in one place.
+
+
+---
+
+### Example: introduce yourself
+
+Benefits of using function:
+
+- we added a new custom functionality to JavaScript
+- by wrapping some code in a function, we removed all the repetitions from the first example
+- to easily repeat set of instructions
+- we hid the "complexities"
+
+---
+
+### Hide complexities
+
+We've used `console.log()` pretty often during those weeks.
+
+Do you have any idea of how it works?
+
+Neither do I.
+<!-- .element: class="fragment" -->
+
+We are still able to use it without any problem.
+<!-- .element: class="fragment" -->
+
+---
+
+### Hide complexities
+
+In real life we don't have to know how things work in order to use them:
+
+- turning on a car
+- using a computer
+- making a phone call
+
+We perform a simple operation (pressing a button, turning a key) without needing to worry about what happens behind the scenes. 
+
+The inner complexities are hidden from us. To an extent, it's what happens with functions.
+
+---
+
+### Avoid repetitions 
+
+The biggest benefit is probably the ability to group many lines of code together and easily execute them as many times as we want without having to write the whole thing again.
+
+```js
+function maximumOf3(value1, value2, value3) {
```

Comment:
ok, in real world, I'd use `maximumOf2(maximumOf2(value1, value2), value3);` :)

Reply:
I'm still trying to avoid putting a function directly inside of console.log() 😅